docs/user-guide/a1-06-spirv-target-specifics.md

```diff
@@ -0,0 +1,316 @@
+`std430` requires a natural alignment, which means that the required alignment is same as its size.
+
+
+Matrix type translation
+-----------------------
+
+When targeting SPIR-V, Slang assumes Row-Major matrix, which is same behavior as HLSL/DXC.
+You can change the behavior to Column-major with a compiler option, "-matrix-layout-column-major".
+
+One of frequently asked questions is that "GLSL/OpenGL uses Column-major and why Slang uses Row-major for targeting SPIR-V?"
+GLSL specification says that "Initialization of matrix values is done with constructors in column-major order", and also says "(In Matrix,) two subscripts select a column and then a row."
+Although it is considered "Column-major", when we access the matrix as 2-dimensional array, the first subscript selects "column" and not "row".
+This is effectively same to a Row-major system when the first subscript selects row and the second subscript selects column.
+
+The actual difference is on the fact that the "memory layout" is in Column-major when OpenGL sends the matrix data to shader.
+When the shader reads the data from the memory, it needs to know how to handle.
+That's why you can specify `row_major` and `column_major` for uniform parameters as their `layout` qualifier in GLSL.
+By default, OpenGL sends the matrix data in Column-major layout, but it is up-to the application to decide how to layout the matrix data.
+
+For more detailed explanation, please checkout another document [a1-01-matrix-layout](a1-01-matrix-layout.md)
+
+
+Legalization (Need to use a more user friendly word)
+------------
+
+TODO: We should also have sections to talk about type legalization and mention all the cases that we enable with our legalization passes.
+
+TODO: For example, we allow using resource types in structs, we allow functions that return resource types as return type or out parameter, as long as things are statically resolvable. We allow functions that return arrays, we allow putting scalar/vector/matrix/array types directly as element type of a constant buffer or structured buffers. Make sure to go over our legalization passes to build a list of allowed things that people are unsure whether it is allowed.
+
+
+Tessellation
+------------
+
+In HLSL and Slang, Hull shader requires two functions: a Hull shader and patch function.
+A typical example of a Hull shader will look like the following.
+```
+// Hull Shader (HS)
+[domain("quad")]
+[patchconstantfunc("constants")]
+HS_OUT main(InputPatch<VS_OUT, 4> patch, uint i : SV_OutputControlPointID)
+{
+  ...
+}
+HSC_OUT constants(InputPatch<VS_OUT, 4> patch)
+{
+  ...
+}
+```
+
+When targeting SPIR-V, the patch function is merged as a part of the Hull shader, because GLSL nor SPIR-V differentiates them like how HLSL does.
+
+As an example, a Hull shader will be emitted as following,
+```
+void main() {
+    // Set tessellation levels
```

Comment:
we will first call the `constants` function, then writes to the outputs on thread 0, then a barrier, then run the `main` function.

Reply:
I omitted the barrier and other parts for the sake of simplicity.
But I guess I shouldn't have.
I will rewrite the code including what you described.

Reply:
I think the correct order is the following,
```
void main() {
    ...
    main(patch, gl_InvocationID);
    barrier();
    if (gl_InvocationID == 0)
    {
        constants(path);
    }
}
```
This matches to how Slangc currently emits them when tested with "tests/spirv/tessellation.slang"
It is implemented in a function, `invokePathConstantFuncInHullShader` [Link](https://github.com/shader-slang/slang/blob/f0ba756c2f982aac8095ff0928d048fc97548315/source/slang/slang-ir-glsl-legalize.cpp#L1049).